File: Xit/XTRepository.m
#import "XTRepository.h"
#import "XTRepository+Commands.h"
#import "XTConstants.h"
#import "Xit-Swift.h"
#import <ObjectiveGit/ObjectiveGit.h>

NSString *XTErrorOutputKey = @"output";
NSString *XTErrorArgsKey = @"args";
NSString *XTPathsKey = @"paths";

NSString *XTErrorDomainXit = @"Xit", *XTErrorDomainGit = @"git";

NSString * const XTRepositoryChangedNotification = @"RepoChanged";
NSString * const XTRepositoryRefsChangedNotification = @"RefsChanged";
NSString * const XTRepositoryIndexChangedNotification = @"IndexChanged";
NSString * const XTRepositoryHeadChangedNotification = @"HeadChanged";


@interface XTRepository ()

@property(readwrite) XTRepositoryWatcher *watcher;
@property(readwrite) BOOL isShutDown;
@property(readwrite) XTConfig *config;

@end

@interface XTRepository (CurrentBranch)

- (NSString*)calculateCurrentBranch;

@end


@implementation XTRepository

+ (NSString *)gitPath
{
  NSArray *paths = @[ @"/usr/bin/git", @"/usr/local/git/bin/git" ];

  for (NSString *path in paths) {
    if ([[NSFileManager defaultManager] fileExistsAtPath:path])
      return path;
  }
  return nil;
}

- (nullable instancetype)initWithURL:(NSURL *)url
{
  self = [super init];
  if (self != nil) {
    NSError *error = nil;

    _gtRepo = [[GTRepository alloc] initWithURL:url error:&error];
    if (error != nil)
      NSLog(@"%@", error.description);
    _gitCMD = [self.class gitPath];
    _repoURL = url;
    NSMutableString *qName =
        [NSMutableString stringWithString:@"com.xit.queue."];
    [qName appendString:url.path];
    _queue = dispatch_queue_create(
        [qName cStringUsingEncoding:NSASCIIStringEncoding],
        DISPATCH_QUEUE_SERIAL);
    _diffCache = [[NSCache alloc] init];
    
    if (_gtRepo != nil) {
      self.watcher = [[XTRepositoryWatcher alloc] initWithRepository:self];
      self.config = [[XTConfig alloc] initWithRepository:self];
      
      [[NSNotificationCenter defaultCenter]
          addObserverForName:XTRepositoryRefsChangedNotification
                      object:self
                       queue:nil
                  usingBlock:^(NSNotification * _Nonnull note) {
        NSString *newBranch = [self calculateCurrentBranch];
        
        if (![_cachedBranch isEqualToString:newBranch]) {
          [self willChangeValueForKey:@"currentBranch"];
          _cachedBranch = newBranch;
          [self didChangeValueForKey:@"currentBranch"];
        }
      }];
    }
  }

  return self;
}

- (void)dealloc
{
  [self.watcher stop];
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (NSURL*)gitDirectoryURL
{
  return _gtRepo.gitDirectoryURL;
}

- (BOOL)executeWritingBlock:(BOOL (^)())block;
{
  BOOL result = NO;

  @synchronized(self) {
    if (self.isWriting)
      return NO;
    self.isWriting = YES;
    result = block();
    self.isWriting = NO;
  }
  return result;
}

- (void)executeOffMainThread:(void (^)())block
{
  if ([NSThread isMainThread]) {
    if (!self.isShutDown)
      dispatch_async(_queue, block);
  } else {
    block();
  }
}

// Make sure KVO notifications happen on the main thread
- (void)updateIsWriting:(BOOL)writing
{
  if (writing == self.isWriting)
    return;
  
  if ([NSThread isMainThread])
    self.isWriting = writing;
  else
    dispatch_sync(dispatch_get_main_queue(), ^{
      self.isWriting = writing;
    });
}

- (void)shutDown
{
  self.isShutDown = YES;
}

- (NSData *)executeGitWithArgs:(NSArray *)args
                        writes:(BOOL)writes
                         error:(NSError **)error
{
  return [self executeGitWithArgs:args
                        withStdIn:nil
                           writes:writes
                            error:error];
}

- (NSData *)executeGitWithArgs:(NSArray *)args
                     withStdIn:(NSString *)stdIn
                        writes:(BOOL)writes
                         error:(NSError **)error
{
  if (_repoURL == nil)
    return nil;
  if (error != NULL)
    *error = nil;

  @synchronized(self) {
    if (writes && self.isWriting) {
      if (error != NULL)
        *error = [NSError errorWithDomain:XTErrorDomainXit
                                     code:XTErrorWriteLock
                                 userInfo:nil];
      return nil;
    }
    [self updateIsWriting:YES];
    NSLog(@"****command = git %@", [args componentsJoinedByString:@" "]);
    NSTask *task = [[NSTask alloc] init];
    [[NSNotificationCenter defaultCenter] postNotificationName:XTTaskStartedNotification object:self];
    task.currentDirectoryPath = _repoURL.path;
    task.launchPath = _gitCMD;
    task.arguments = args;

    if (stdIn != nil) {
#if 0
      NSLog(@"**** stdin = %lu", stdIn.length);
#else
      NSLog(@"**** stdin = %lu\n%@", stdIn.length, stdIn);
#endif
      NSPipe *stdInPipe = [NSPipe pipe];
      [stdInPipe.fileHandleForWriting
          writeData:[stdIn dataUsingEncoding:NSUTF8StringEncoding]];
      [stdInPipe.fileHandleForWriting closeFile];
      task.standardInput = stdInPipe;
    }

    NSPipe *pipe = [NSPipe pipe];
    NSPipe *errorPipe = [NSPipe pipe];

    task.standardOutput = pipe;
    task.standardError = errorPipe;

    NSLog(@"task.currentDirectoryPath=%@", task.currentDirectoryPath);
    [task launch];
    NSData *output = [pipe.fileHandleForReading readDataToEndOfFile];
    [task waitUntilExit];

    const int status = task.terminationStatus;
    NSLog(@"**** status = %d", status);

    if (status != 0) {
      NSString *string =
          [[NSString alloc] initWithData:output encoding:NSUTF8StringEncoding];
      NSData *errorOutput =
          [errorPipe.fileHandleForReading readDataToEndOfFile];
      NSString *errorString =
          [[NSString alloc] initWithData:errorOutput encoding:NSUTF8StringEncoding];
      NSLog(@"**** output = %@", string);
      NSLog(@"**** error = %@", errorString);
      if (error != NULL) {
        NSDictionary *info =
            @{ XTErrorOutputKey:string,
               XTErrorArgsKey:[args componentsJoinedByString:@" "] };

        *error = [NSError errorWithDomain:@"git" code:status userInfo:info];
      }
      output = nil;
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:XTTaskEndedNotification object:self];
    [self updateIsWriting:NO];
    return output;
  }
}

- (BOOL)hasHeadReference
{
  NSError *error = nil;

  return [_gtRepo headReferenceWithError:&error] != nil;
}

- (NSString *)parseSymbolicReference:(NSString *)reference
{
  NSError *error = nil;
  GTReference *gtRef = [_gtRepo lookUpReferenceWithName:reference error:&error];

  if (error != nil)
    return nil;

  id unresolved = gtRef.unresolvedTarget;

  if (![unresolved isKindOfClass:[GTReference class]])
    return reference;
  return [(GTReference*)unresolved name];
}

/// Returns kEmptyTreeHash if the repository is empty, otherwise "HEAD"
- (NSString *)parentTree
{
  return [self hasHeadReference] ? @"HEAD" : kEmptyTreeHash;
}

- (NSString *)shaForRef:(NSString *)ref
{
  if (ref == nil)
    return nil;

  NSError *error = nil;
  GTObject *object = [_gtRepo lookUpObjectByRevParse:ref error:&error];

  if (error != nil)
    return nil;
  return object.SHA;
}

- (NSString *)headRef
{
  @synchronized(self) {
    if (_cachedHeadRef == nil) {
      NSString *head = [self parseSymbolicReference:@"HEAD"];

      if ([head hasPrefix:@"refs/heads/"])
        _cachedHeadRef = head;
      else
        _cachedHeadRef = @"HEAD";

      _cachedHeadSHA = [self shaForRef:_cachedHeadRef];
    }
  }
  return _cachedHeadRef;
}

- (NSString*)headSHA
{
  return [self shaForRef:[self headRef]];
}

- (NSArray<NSString*>*)remoteNames
{
  NSArray<NSString*> *result = [_gtRepo remoteNamesWithError:NULL];
  
  if (result == nil)
    result = @[];
  return result;
}

- (NSData*)contentsOfFile:(NSString*)filePath
                 atCommit:(NSString*)commitSHA
                    error:(NSError**)error
{
  GTCommit *commit = [self.gtRepo lookUpObjectBySHA:commitSHA error:error];

  if (![commit isKindOfClass:[GTCommit class]]) {
    if (*error == nil)
      *error = [NSError errorWithDomain:XTErrorDomainXit
                                   code:XTErrorUnexpectedObject
                               userInfo:nil];
    return nil;
  }

  GTTree *tree = commit.tree;
  GTTreeEntry *entry = [tree entryWithPath:filePath error:error];
  GTBlob *blob = (GTBlob*)[entry GTObject:error];

  if (![blob isKindOfClass:[GTBlob class]]) {
    if (*error == nil)
      *error = [NSError errorWithDomain:XTErrorDomainXit
                                   code:XTErrorUnexpectedObject
                               userInfo:nil];
    return nil;
  }
  return blob.data;
}

- (NSData*)contentsOfStagedFile:(NSString*)filePath
                          error:(NSError**)error
{
  GTIndex *index = [self.gtRepo indexWithError:error];
  
  if (index == nil)
    return nil;
  // GTRepository returns any cached index object it had, so it may need
  // to be reloaded.
  if (![index refresh:error])
    return nil;
  
  GTIndexEntry *entry = [index entryWithPath:filePath error:error];
  GTBlob *blob = (GTBlob*)[entry GTObject:error];
  
  if (![blob isKindOfClass:[GTBlob class]]) {
    if (*error == nil)
      *error = [NSError errorWithDomain:XTErrorDomainXit
                                   code:XTErrorUnexpectedObject
                               userInfo:nil];
    return nil;
  }
  return blob.data;
}

// A convenience method for adding to the default notification center.
- (void)addReloadObserver:(id)observer selector:(SEL)selector
{
  [[NSNotificationCenter defaultCenter]
      addObserver:observer
         selector:selector
             name:XTRepositoryChangedNotification
           object:self];
}

@end

File: Xit/XTRepository.swift
import Foundation


extension NSNotification.Name
{
  /// Some change has been detected in the repository.
  static let XTRepositoryChanged =
      NSNotification.Name(rawValue: "XTRepositoryChanged")
  /// The repository's config file has changed.
  static let XTRepositoryConfigChanged =
      NSNotification.Name(rawValue: "XTRepositoryConfigChanged")
  /// The head reference (current branch) has changed.
  static let XTRepositoryHeadChanged =
      NSNotification.Name(rawValue: "XTRepositoryHeadChanged")
  /// The repository's index has changed.
  static let XTRepositoryIndexChanged =
      NSNotification.Name(rawValue: "XTRepositoryIndexChanged")
  /// The repository's refs have changed.
  static let XTRepositoryRefsChanged =
      NSNotification.Name(rawValue: "XTRepositoryRefsChanged")
  /// A file in the workspace has changed.
  static let XTRepositoryWorkspaceChanged =
      NSNotification.Name(rawValue: "XTRepositoryWorkspaceChanged")
  /// The stash log has changed.
  static let XTRepositoryStashChanged =
      NSNotification.Name(rawValue: "XTRepositoryStashChanged")
}


/// Stores a repo reference for C callbacks
struct CallbackPayload { let repo: XTRepository }

let kEmptyTreeHash = "4b825dc642cb6eb9a060e54bf8d69288fbee4904"
let XTPathsKey = "paths"
let XTErrorOutputKey = "output"
let XTErrorArgsKey = "args"

public class XTRepository: NSObject
{
  private(set) var gtRepo: GTRepository
  @objc public let repoURL: URL
  let gitCMD: String
  @objc let queue: TaskQueue
  let mutex = Mutex()
  var refsIndex = [String: [String]]()
  fileprivate(set) var isWriting = false
  fileprivate var executing = false
  
  fileprivate(set) var cachedHeadRef, cachedHeadSHA, cachedBranch: String?
  private var _cachedStagedChanges, _cachedAmendChanges,
              _cachedUnstagedChanges: [FileChange]?
  var cachedStagedChanges: [FileChange]?
  {
    get { return mutex.withLock { _cachedStagedChanges } }
    set { mutex.withLock { _cachedStagedChanges = newValue } }
  }
  var cachedAmendChanges: [FileChange]?
  {
    get { return mutex.withLock { _cachedAmendChanges } }
    set { mutex.withLock { _cachedAmendChanges = newValue } }
  }
  var cachedUnstagedChanges: [FileChange]?
  {
    get { return mutex.withLock { _cachedUnstagedChanges } }
    set { mutex.withLock { _cachedUnstagedChanges = newValue } }
  }

  let diffCache = Cache<String, Diff>(maxSize: 50)
  fileprivate var repoWatcher: XTRepositoryWatcher! = nil
  fileprivate var workspaceWatcher: WorkspaceWatcher! = nil
  private(set) var config: XTConfig! = nil
  
  var gitRepo: OpaquePointer { return gtRepo.git_repository() }
  
  var gitDirectoryPath: String
  {
    guard let path = git_repository_path(gitRepo)
    else { return "" }
    
    return String(cString: path)
  }
  
  var gitDirectoryURL: URL
  {
    return URL(fileURLWithPath: gitDirectoryPath)
  }
  
  static func gitPath() -> String?
  {
    let paths = ["/usr/bin/git", "/usr/local/git/bin/git"]
    
    return paths.first(where: { FileManager.default.fileExists(atPath: $0) })
  }
  
  @objc(initWithURL:)
  init?(url: URL)
  {
    guard let gitCMD = XTRepository.gitPath(),
          let gtRepo = try? GTRepository(url: url)
    else { return nil }
    
    self.repoURL = url
    self.gitCMD = gitCMD
    self.gtRepo = gtRepo
    
    self.queue = TaskQueue(id: "com.uncommonplace.xit.\(url.path)")
    
    super.init()
    
    postInit()
  }
  
  @objc(initEmptyWithURL:)
  init?(emptyURL url: URL)
  {
    guard let gitCMD = XTRepository.gitPath(),
          let gtRepo = try? GTRepository.initializeEmpty(atFileURL: url,
                                                         options: nil)
    else { return nil }
    
    self.repoURL = url
    self.gitCMD = gitCMD
    self.gtRepo = gtRepo
    self.queue = TaskQueue(id: "com.uncommonplace.xit.\(url.path)")
    
    super.init()
    
    postInit()
  }
  
  private func postInit()
  {
    self.repoWatcher = XTRepositoryWatcher(repository: self)
    self.workspaceWatcher = WorkspaceWatcher(repository: self)
    self.config = XTConfig(config: GitConfig(repository: gitRepo))
  }
  
  deinit
  {
    repoWatcher.stop()
    workspaceWatcher.stop()
    NotificationCenter.default.removeObserver(self)
  }
  
  func updateIsWriting(_ writing: Bool)
  {
    guard writing != isWriting
    else { return }
    
    mutex.withLock {
      isWriting = writing
    }
  }
  
  func performWriting(_ block: (() throws -> Void)) throws
  {
    objc_sync_enter(self)
    defer { objc_sync_exit(self) }
    
    if isWriting {
      throw Error.alreadyWriting
    }
    updateIsWriting(true)
    defer {
      updateIsWriting(false)
    }
    try block()
  }
  
  func clearCachedBranch()
  {
    mutex.withLock {
      cachedBranch = nil
    }
  }
  
  func refsChanged()
  {
    // In theory the two separate locks could result in cachedBranch being wrong
    // but that would only happen if this function was called on two different
    // threads and one of them found that the branch had just changed again.
    // Not likely.
    guard let newBranch = calculateCurrentBranch(),
          mutex.withLock({ newBranch != cachedBranch })
    else { return }
    
    willChangeValue(forKey: "currentBranch")
    mutex.withLock {
      cachedBranch = newBranch
    }
    didChangeValue(forKey: "currentBranch")
  }
  
  func recalculateHead()
  {
    guard let headReference = self.headReference
    else { return }
    
    switch headReference.type {
      case .symbolic:
        cachedHeadRef = headReference.symbolicTargetName
      case .OID:
        cachedHeadRef = headReference.name
      default:
        break
    }
    cachedHeadSHA = sha(forRef: headReference.name)
  }
  
  func invalidateIndex()
  {
    mutex.withLock {
      cachedStagedChanges = nil
      cachedAmendChanges = nil
      cachedUnstagedChanges = nil
    }
  }
  
  func writing(_ block: () -> Bool) -> Bool
  {
    objc_sync_enter(self)
    defer {
      objc_sync_exit(self)
    }
    
    guard !isWriting
    else { return false }
    
    isWriting = true
    defer {
      isWriting = false
    }
    return block()
  }
  
  func executeGit(args: [String],
                  stdIn: String?,
                  writes: Bool) throws -> Data
  {
    return try executeGit(args: args,
                          stdInData: stdIn?.data(using: .utf8),
                          writes: writes)
  }
  
  func executeGit(args: [String], writes: Bool) throws -> Data
  {
    return try executeGit(args: args, stdInData: nil, writes: writes)
  }
  
  func executeGit(args: [String],
                  stdInData: Data?,
                  writes: Bool) throws -> Data
  {
    guard FileManager.default.fileExists(atPath: repoURL.path)
    else {
      throw NSError(domain: NSCocoaErrorDomain, code: NSFileNoSuchFileError,
                    userInfo: nil)
    }
    
    objc_sync_enter(self)
    defer {
      objc_sync_exit(self)
    }
    if writes && isWriting {
      throw Error.alreadyWriting
    }
    updateIsWriting(writes)
    defer {
      updateIsWriting(false)
    }
    executing = true
    defer { executing = false }
    NSLog("*** command = git \(args.joined(separator: " "))")
    
    let task = Process()
    
    task.currentDirectoryPath = repoURL.path
    task.launchPath = gitCMD
    task.arguments = args
    
    // Large files have to be chunked or else FileHandle.write() hangs
    let chunkSize = 10*1024

    if let data = stdInData {
      let stdInPipe = Pipe()
      
      if data.count <= chunkSize {
        stdInPipe.fileHandleForWriting.write(data)
        stdInPipe.fileHandleForWriting.closeFile()
      }
      task.standardInput = stdInPipe
    }
    
    let pipe = Pipe()
    let errorPipe = Pipe()
    
    task.standardOutput = pipe
    task.standardError = errorPipe
    try task.throwingLaunch()
    
    if let data = stdInData,
       data.count > chunkSize,
       let handle = (task.standardInput as? Pipe)?.fileHandleForWriting {
      for chunkIndex in 0...(data.count/chunkSize) {
        let chunkStart = chunkIndex * chunkSize
        let chunkEnd = min(chunkStart + chunkSize, data.count)
        let subData = data.subdata(in: chunkStart..<chunkEnd)
        
        handle.write(subData)
      }
      handle.closeFile()
    }
    
    let output = pipe.fileHandleForReading.readDataToEndOfFile()
    
    task.waitUntilExit()
    
    guard task.terminationStatus == 0
    else {
      let string = String(data: output, encoding: .utf8) ?? "-"
      let errorOutput = errorPipe.fileHandleForReading.readDataToEndOfFile()
      let errorString = String(data: errorOutput, encoding: .utf8) ?? "-"
      
      NSLog("**** output = \(string)")
      NSLog("**** error = \(errorString)")
      throw NSError(domain: XTErrorDomainGit, code: Int(task.terminationStatus),
                    userInfo: [XTErrorOutputKey: string,
                               XTErrorArgsKey: args.joined(separator: " ")])
    }
    
    return output
  }
}

// For testing
internal func setRepoWriting(_ repo: XTRepository, _ writing: Bool)
{
  repo.isWriting = writing
}

extension XTRepository: CommitStorage
{
  public func oid(forSHA sha: String) -> OID?
  {
    return GitOID(sha: sha)
  }
  
  public func commit(forSHA sha: String) -> Commit?
  {
    return XTCommit(sha: sha, repository: self)
  }
  
  public func commit(forOID oid: OID) -> Commit?
  {
    return XTCommit(oid: oid, repository: gitRepo)
  }
  
  public func walker() -> RevWalk?
  {
    return GitRevWalk(repository: gitRepo)
  }
}

extension XTRepository
{
  /// Returns true if the path is ignored according to the repository's
  /// ignore rules.
  func isIgnored(path: String) -> Bool
  {
    let ignored = UnsafeMutablePointer<Int32>.allocate(capacity: 1)
    let result = git_ignore_path_is_ignored(ignored, gitRepo, path)

    return (result == 0) && (ignored.pointee != 0)
  }
  
  /// Returns the unstaged and staged status of the given file.
  func status(file: String) throws -> (DeltaStatus, DeltaStatus)
  {
    let statusFlags = UnsafeMutablePointer<UInt32>.allocate(capacity: 1)
    let result = git_status_file(statusFlags, gitRepo, file)
    
    if result != 0 {
      throw NSError.git_error(for: result)
    }
    
    let flags = git_status_t(statusFlags.pointee)
    var unstagedChange = DeltaStatus.unmodified
    var stagedChange = DeltaStatus.unmodified
    
    switch flags {
      case _ where flags.test(GIT_STATUS_WT_NEW):
        unstagedChange = .untracked
      case _ where flags.test(GIT_STATUS_WT_MODIFIED),
           _ where flags.test(GIT_STATUS_WT_TYPECHANGE):
        unstagedChange = .modified
      case _ where flags.test(GIT_STATUS_WT_DELETED):
        unstagedChange = .deleted
      case _ where flags.test(GIT_STATUS_WT_RENAMED):
        unstagedChange = .renamed
      case _ where flags.test(GIT_STATUS_IGNORED):
        unstagedChange = .ignored
      case _ where flags.test(GIT_STATUS_CONFLICTED):
        unstagedChange = .conflict
      // ignoring GIT_STATUS_WT_UNREADABLE
      default:
        break
    }
    
    switch flags {
      case _ where flags.test(GIT_STATUS_INDEX_NEW):
        stagedChange = .added
      case _ where flags.test(GIT_STATUS_INDEX_MODIFIED),
           _ where flags.test(GIT_STATUS_WT_TYPECHANGE):
        stagedChange = .modified
      case _ where flags.test(GIT_STATUS_INDEX_DELETED):
        stagedChange = .deleted
      case _ where flags.test(GIT_STATUS_INDEX_RENAMED):
        stagedChange = .renamed
      default:
        break
    }
    
    return (unstagedChange, stagedChange)
  }
  
  func graphBetween(local: OID, upstream: OID) -> (ahead: Int,
                                                   behind: Int)?
  {
    guard let local = local as? GitOID,
          let upstream = upstream as? GitOID
    else { return nil }
    let ahead = UnsafeMutablePointer<Int>.allocate(capacity: 1)
    let behind = UnsafeMutablePointer<Int>.allocate(capacity: 1)
    
    if git_graph_ahead_behind(ahead, behind, gitRepo,
                              local.unsafeOID(), upstream.unsafeOID()) == 0 {
      return (ahead.pointee, behind.pointee)
    }
    else {
      return nil
    }
  }
  
  public func graphBetween(localBranch: LocalBranch,
                           upstreamBranch: RemoteBranch) ->(ahead: Int,
                                                            behind: Int)?
  {
    if let localOID = localBranch.oid,
       let upstreamOID = upstreamBranch.oid {
      return graphBetween(local: localOID, upstream: upstreamOID)
    }
    else {
      return nil
    }
  }
}

extension NSNotification.Name
{
  static let XTRepositoryRefLogChanged =
      NSNotification.Name("XTRepositoryRefLogChanged")
}
